pkg/nodeinfo/attributes.go

```
@@ -0,0 +1,99 @@
+const (
+	nodeLabelOSName        = "feature.node.kubernetes.io/system-os_release.ID"
+	nodeLabelOSVer         = "feature.node.kubernetes.io/system-os_release.VERSION_ID"
+	nodeLabelKernelVerFull = "feature.node.kubernetes.io/kernel-version.full"
+	nodeLabelHostname      = "kubernetes.io/hostname"
+	nodeLabelCPUArch       = "kubernetes.io/arch"
+	nodeLabelMlnxNIC       = "feature.node.kubernetes.io/pci-15b3.present"
+)
+
+type AttributeType int
+
+const (
+	AttrTypeOS = iota
+	AttrTypeKernel
+	AttrTypeHostname
+	AttrTypeCPUArch
+)
+
+// NodeAttributes provides attributes of a specific node
+type NodeAttributes struct {
+	// Node Name
+	Name string
+	// Node Attributes
+	Attributes map[AttributeType]string
+}
+
+// fromLabels adds a new attribute of type attrT to NodeAttributes by joining value of selectedLabels
+// using delim delimiter
+//nolint:unparam
+func (a *NodeAttributes) fromLabels(
+	attrT AttributeType, nodeLabels map[string]string, selectedLabels []string, delim string) error {
+	var attrVal string
+
+	//nolint:prealloc
+	var labelVals []string
+	for _, sLabel := range selectedLabels {
+		lVal, ok := nodeLabels[sLabel]
+		if !ok {
+			return errors.Errorf("cannot create node attribute, missing label: %s", sLabel)
+		}
+		labelVals = append(labelVals, lVal)
+	}
+	attrVal = strings.Join(labelVals, delim)
+
+	// Note: attrVal may be empty, this could indicate a binary attribute which relies on key existence
+	a.Attributes[attrT] = attrVal
+	return nil
+}
+
+// newAttributesFromNode creates a new NodeAttributes
+func newAttributesFromNode(node *corev1.Node) NodeAttributes {
+	attr := NodeAttributes{
+		Name:       node.GetName(),
+		Attributes: make(map[AttributeType]string),
+	}
+	var err error
+
+	nLabels := node.GetLabels()
+	// OS name
+	err = attr.fromLabels(AttrTypeOS, nLabels, []string{nodeLabelOSName, nodeLabelOSVer}, "")
+	if err != nil {
+		log.V(consts.LogDebugLevel).Info("Warning:", "cannot create NodeAttributes, %v", err)
```

Info or Warning?

---

The logger only got INFO when using V style logging

without V style logging there is either info or error.

IMO at this level Failing to create an attribute should be considered a warning.
Upper layers should still check and fail if an attribute is missing.

Thinking on this, i marked the V as "debug" which is not really good,
i will define a more granular set of Logging verbosity consts as this should definetaly be visible